src/app/reporting/Engine.cpp

```diff
@@ -336,6 +367,66 @@ void Engine::Run()
         mCurReadHandlerIdx = (mCurReadHandlerIdx + 1) % CHIP_IM_MAX_NUM_READ_HANDLER;
         readHandler        = imEngine->mReadHandlers + mCurReadHandlerIdx;
     }
+
+    bool allReadClean = true;
+    for (auto & handler : InteractionModelEngine::GetInstance()->mReadHandlers)
+    {
+        UpdateReadHandlerDirty(handler);
+        if (handler.IsDirty())
+        {
+            allReadClean = false;
+            break;
+        }
+    }
+
+    if (allReadClean)
+    {
+        InteractionModelEngine::GetInstance()->ReleaseClusterInfoList(mpChangePathList);
+    }
+}
+
+CHIP_ERROR Engine::SetDirty(ClusterInfo & aClusterInfo)
+{
+    for (auto & handler : InteractionModelEngine::GetInstance()->mReadHandlers)
+    {
+        if (handler.IsSubscriptionType() && (handler.IsGeneratingReports() || handler.IsAwaitingReportResponse()))
+        {
+            handler.SetDirty();
+        }
+    }
+    if (!InteractionModelEngine::GetInstance()->MergeOverlappedAttributePath(mpChangePathList, aClusterInfo) &&
+        InteractionModelEngine::GetInstance()->IsOverlappedAttributePath(aClusterInfo))
+    {
+        ReturnLogErrorOnFailure(InteractionModelEngine::GetInstance()->PushFront(mpChangePathList, aClusterInfo));
+    }
+    return CHIP_NO_ERROR;
+}
+
+void Engine::UpdateReadHandlerDirty(ReadHandler & aReadHandler)
```

Comment:
Please add a comment about what this is doing in the header file.

Reply:
Done, thanks